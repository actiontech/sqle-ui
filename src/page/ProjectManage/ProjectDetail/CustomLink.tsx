import { Link } from 'react-router-dom';
import { CustomLinkProps, ProjectDetailCustomLinkState } from '.';

/**
 *
 * @param param0
 * @returns
 */
const CustomLink: React.FC<CustomLinkProps> = ({
  to,
  projectName,
  search,
  children,
  ...props
}) => {
  return (
    <Link<ProjectDetailCustomLinkState>
      {...props}
      to={{ pathname: to, search, state: { projectName } }}
    >
      {children}
    </Link>
  );
};

export default CustomLink;
