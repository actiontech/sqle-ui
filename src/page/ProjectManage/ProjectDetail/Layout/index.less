.project-detail-wrapper {
  min-height: calc(100vh - 50px);
}
