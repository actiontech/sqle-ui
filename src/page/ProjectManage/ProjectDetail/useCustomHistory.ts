import { useHistory } from 'react-router-dom';
import { ProjectDetailCustomLinkState } from '.';

const useCustomHistory = () => {
  const history = useHistory<ProjectDetailCustomLinkState>();
  const push = (path: string, projectName: string) => {
    history.push(path, { projectName });
  };

  const replace = (path: string, projectName: string) => {
    history.replace(path, { projectName });
  };

  return {
    ...history,
    push,
    replace,
  };
};

export default useCustomHistory;
