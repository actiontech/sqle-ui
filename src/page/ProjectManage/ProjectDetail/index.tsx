import ProjectDetail from './ProjectDetail';

export type ProjectDetailLocationStateType = {
  projectName: string;
};

export default ProjectDetail;
