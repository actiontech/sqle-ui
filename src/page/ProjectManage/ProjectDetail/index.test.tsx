import ProjectDetail, { useRecentlyOpenedProjects } from '.';
import { useLocation, useParams } from 'react-router-dom';
import {
  mockBindProjects,
  mockManagementPermissions,
} from '../../../hooks/useCurrentUser/index.test';
import { render, screen, act as reactAct } from '@testing-library/react';
import { renderWithRouter } from '../../../testUtils/customRender';
import {
  mockUseAuditPlanTypes,
  mockUseInstance,
  mockUseUsername,
  resolveErrorThreeSecond,
  resolveThreeSecond,
} from '../../../testUtils/mockRequest';
import user from '../../../api/user';
import {
  mockGetProjectDetail,
  mockGetProjectStatistics,
} from '../__test__/utils';
import { mockUseStyle } from '../../../testUtils/mockStyle';
import EventEmitter from '../../../utils/EventEmitter';
import EmitterKey from '../../../data/EmitterKey';
import StorageKey from '../../../data/StorageKey';
import { SystemRole } from '../../../data/common';
import { act, renderHook } from '@testing-library/react-hooks';
import useNavigate from '../../../hooks/useNavigate';
import { useDispatch, useSelector } from 'react-redux';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: jest.fn(),
  useLocation: jest.fn(),
}));

jest.mock('react-redux', () => {
  return {
    ...jest.requireActual('react-redux'),
    useSelector: jest.fn(),
    useDispatch: jest.fn(),
  };
});

jest.mock('../../../hooks/useNavigate', () => jest.fn());

const projectName = mockBindProjects[0].project_name;

describe('test ProjectManage/ProjectDetail', () => {
  const useParamsMock: jest.Mock = useParams as jest.Mock;
  let getUserSpy: jest.SpyInstance;
  const dispatchSpy = jest.fn();
  let getProjectDetailSpy: jest.SpyInstance;
  const useLocationMock: jest.Mock = useLocation as jest.Mock;
  const useHistoryMock: jest.Mock = useNavigate as jest.Mock;
  const navigateSpy = jest.fn();
  beforeEach(() => {
    getUserSpy = jest.spyOn(user, 'getCurrentUserV1');
    getUserSpy.mockImplementation(() =>
      resolveThreeSecond({
        user_name: 'test',
        is_admin: '',
        bind_projects: mockBindProjects,
      })
    );
    getProjectDetailSpy = mockGetProjectDetail();
    (useDispatch as jest.Mock).mockImplementation(() => dispatchSpy);

    mockUseAuditPlanTypes();
    mockUseUsername();
    mockUseInstance();
    mockUseStyle();
    useLocationMock.mockReturnValue({
      pathname: '/project/default/overview',
      search: '',
      hash: '',
      state: null,
      key: '5nvxpbdafa',
    });
    useHistoryMock.mockImplementation(() => navigateSpy);
    useParamsMock.mockReturnValue({ projectName });
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.useRealTimers();
    jest.clearAllTimers();
    useLocationMock.mockRestore();
    useHistoryMock.mockRestore();
  });
  test('should render tips for unbound project', async () => {
    getUserSpy.mockImplementation(() =>
      resolveThreeSecond({
        user_name: 'test',
        is_admin: '',
        bind_projects: [],
      })
    );
    expect(getUserSpy).toBeCalledTimes(0);

    render(<ProjectDetail />);
    expect(getUserSpy).toBeCalledTimes(1);
    expect(getProjectDetailSpy).toBeCalledTimes(1);
    await reactAct(async () => jest.advanceTimersByTime(3000));
    expect(
      screen.getByText('projectManage.projectDetail.unboundProjectTips')
    ).toBeInTheDocument();
  });

  test('should clear user info when get current user is wrong', async () => {
    getUserSpy.mockImplementation(() => resolveErrorThreeSecond(undefined));
    expect(dispatchSpy).toBeCalledTimes(0);
    render(<ProjectDetail />);
    await reactAct(async () => jest.advanceTimersByTime(3000));

    expect(dispatchSpy).toBeCalledTimes(5);
    expect(dispatchSpy).nthCalledWith(1, {
      payload: false,
      type: 'projectManage/updateProjectStatus',
    });
    expect(dispatchSpy).nthCalledWith(2, {
      payload: { bindProjects: [] },
      type: 'user/updateBindProjects',
    });
    expect(dispatchSpy).nthCalledWith(3, {
      payload: { username: '', role: '' },
      type: 'user/updateUser',
    });
    expect(dispatchSpy).nthCalledWith(4, {
      payload: { token: '' },
      type: 'user/updateToken',
    });
    expect(dispatchSpy).nthCalledWith(5, {
      payload: { managementPermissions: [] },
      type: 'user/updateManagementPermissions',
    });
  });

  test('should match snapshot when bound project exists', async () => {
    const { container } = renderWithRouter(<ProjectDetail />);
    await reactAct(async () => jest.advanceTimersByTime(3000));
    await reactAct(async () => jest.advanceTimersByTime(0));

    expect(container).toMatchSnapshot();
    expect(getProjectDetailSpy).toBeCalledTimes(1);
  });

  test('should show unavailable when project archived is equal true', async () => {
    mockGetProjectStatistics();
    getProjectDetailSpy.mockImplementation(() =>
      resolveThreeSecond({
        create_time: '2022-01-01',
        create_user_name: 'admin',
        desc: 'desc',
        name: 'project1',
        archived: true,
      })
    );

    const { container } = renderWithRouter(<ProjectDetail />);
    await reactAct(async () => jest.advanceTimersByTime(3000));
    await reactAct(async () => jest.advanceTimersByTime(0));

    expect(
      screen.getByText('(projectManage.projectList.column.unavailable)')
    ).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});

describe('test useRecentlyOpenedProjects', () => {
  const username1 = 'admin';
  const username2 = 'test';

  const bindProjects = [
    {
      is_manager: true,
      project_name: 'default',
    },
    {
      is_manager: false,
      project_name: 'default1',
    },
    {
      is_manager: false,
      project_name: 'default2',
    },
    {
      is_manager: false,
      project_name: 'default3',
    },
    {
      is_manager: false,
      project_name: 'default4',
    },
  ];

  beforeEach(() => {
    jest.useFakeTimers();
    (useSelector as jest.Mock).mockImplementation((e) =>
      e({
        user: {
          username: username1,
          role: SystemRole.admin,
          bindProjects: bindProjects,
          managementPermissions: mockManagementPermissions,
        },
      })
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.useRealTimers();
    jest.clearAllTimers();
  });

  test('should perform as expected with update operation', () => {
    const localStorageSetItemSpy = jest.spyOn(Storage.prototype, 'setItem');
    const { result } = renderHook(() => useRecentlyOpenedProjects());

    expect(localStorageSetItemSpy).toBeCalledTimes(0);

    act(() => {
      result.current.updateRecentlyProject('default1');
    });

    expect(localStorageSetItemSpy).toBeCalledTimes(1);
    expect(localStorageSetItemSpy).toBeCalledWith(
      StorageKey.Project_Catch,
      JSON.stringify({ [username1]: ['default1'] })
    );

    expect(result.current.recentlyProjects).toEqual(['default1']);

    act(() => {
      result.current.updateRecentlyProject('default2');
    });

    expect(localStorageSetItemSpy).toBeCalledTimes(2);
    expect(localStorageSetItemSpy).toBeCalledWith(
      StorageKey.Project_Catch,
      JSON.stringify({ [username1]: ['default2', 'default1'] })
    );

    expect(result.current.recentlyProjects).toEqual(['default2', 'default1']);

    act(() => {
      result.current.updateRecentlyProject('default3');
    });

    expect(localStorageSetItemSpy).toBeCalledTimes(3);
    expect(localStorageSetItemSpy).toBeCalledWith(
      StorageKey.Project_Catch,
      JSON.stringify({ [username1]: ['default3', 'default2', 'default1'] })
    );

    expect(result.current.recentlyProjects).toEqual([
      'default3',
      'default2',
      'default1',
    ]);

    act(() => {
      result.current.updateRecentlyProject('default4');
    });

    expect(localStorageSetItemSpy).toBeCalledTimes(4);
    expect(localStorageSetItemSpy).toBeCalledWith(
      StorageKey.Project_Catch,
      JSON.stringify({ [username1]: ['default4', 'default3', 'default2'] })
    );

    expect(result.current.recentlyProjects).toEqual([
      'default4',
      'default3',
      'default2',
    ]);

    act(() => {
      result.current.updateRecentlyProject('default3');
    });

    expect(localStorageSetItemSpy).toBeCalledTimes(5);
    expect(localStorageSetItemSpy).toBeCalledWith(
      StorageKey.Project_Catch,
      JSON.stringify({ [username1]: ['default3', 'default4', 'default2'] })
    );

    expect(result.current.recentlyProjects).toEqual([
      'default3',
      'default4',
      'default2',
    ]);

    window.localStorage.clear();
  });

  test('should be distinguished projects by users and filtering out projects that do not have permissions', () => {
    const localStorageSetItemSpy = jest.spyOn(Storage.prototype, 'setItem');
    const { result, rerender } = renderHook(() => useRecentlyOpenedProjects());

    expect(localStorageSetItemSpy).toBeCalledTimes(0);

    act(() => {
      result.current.updateRecentlyProject('default1');
    });

    expect(localStorageSetItemSpy).toBeCalledTimes(1);
    expect(localStorageSetItemSpy).toBeCalledWith(
      StorageKey.Project_Catch,
      JSON.stringify({ [username1]: ['default1'] })
    );
    expect(result.current.recentlyProjects).toEqual(['default1']);
    jest.clearAllMocks();

    (useSelector as jest.Mock).mockImplementation((e) =>
      e({
        user: {
          username: username2,
          role: SystemRole.admin,
          bindProjects: bindProjects,
          managementPermissions: mockManagementPermissions,
        },
      })
    );

    rerender();

    act(() => {
      result.current.updateRecentlyProject('default1');
    });

    expect(localStorageSetItemSpy).toBeCalledTimes(1);
    expect(localStorageSetItemSpy).toBeCalledWith(
      StorageKey.Project_Catch,
      JSON.stringify({ [username1]: ['default1'], [username2]: ['default1'] })
    );
    expect(result.current.recentlyProjects).toEqual(['default1']);

    jest.clearAllMocks();

    (useSelector as jest.Mock).mockImplementation((e) =>
      e({
        user: {
          username: username2,
          role: SystemRole.admin,
          bindProjects: [],
          managementPermissions: mockManagementPermissions,
        },
      })
    );

    rerender();
    expect(result.current.recentlyProjects).toEqual([]);
  });

  test('should be an empty array when no update operation is performed', () => {
    const subscribeSpy = jest.spyOn(EventEmitter, 'subscribe');
    const unsubscribeSpy = jest.spyOn(EventEmitter, 'unsubscribe');

    const { result, unmount } = renderHook(() => useRecentlyOpenedProjects());

    expect(result.current.recentlyProjects).toEqual([]);

    expect(unsubscribeSpy).toBeCalledTimes(0);
    expect(subscribeSpy).toBeCalledTimes(1);
    expect(subscribeSpy.mock.calls[0][0]).toBe(
      EmitterKey.Update_Recently_Opened_Projects
    );

    unmount();

    expect(unsubscribeSpy).toBeCalledTimes(1);
    expect(unsubscribeSpy.mock.calls[0][0]).toBe(
      EmitterKey.Update_Recently_Opened_Projects
    );
  });
});
