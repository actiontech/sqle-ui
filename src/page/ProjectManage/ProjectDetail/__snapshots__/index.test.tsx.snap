// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ProjectManage/ProjectDetail should match snapshot when bound project exists 1`] = `
<div>
  <section
    class="ant-layout ant-layout-has-sider project-detail-wrapper"
  >
    <aside
      class="ant-layout-sider ant-layout-sider-dark"
      style="flex: 0 0 200px; max-width: 200px; min-width: 200px; width: 200px;"
    >
      <div
        class="ant-layout-sider-children"
      >
        <h1
          class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line"
          style="color: white; padding: 20px;"
        >
          default
        </h1>
        <ul
          class="ant-menu ant-menu-root ant-menu-inline ant-menu-dark"
          data-menu-list="true"
          role="menu"
          tabindex="0"
        >
          <li
            class="ant-menu-item ant-menu-item-selected"
            data-menu-id="rc-menu-uuid-test-projectOverview"
            role="menuitem"
            style="padding-left: 24px;"
            tabindex="-1"
          >
            <span
              aria-label="project"
              class="anticon anticon-project ant-menu-item-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="project"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M280 752h80c4.4 0 8-3.6 8-8V280c0-4.4-3.6-8-8-8h-80c-4.4 0-8 3.6-8 8v464c0 4.4 3.6 8 8 8zm192-280h80c4.4 0 8-3.6 8-8V280c0-4.4-3.6-8-8-8h-80c-4.4 0-8 3.6-8 8v184c0 4.4 3.6 8 8 8zm192 72h80c4.4 0 8-3.6 8-8V280c0-4.4-3.6-8-8-8h-80c-4.4 0-8 3.6-8 8v256c0 4.4 3.6 8 8 8zm216-432H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
                />
              </svg>
            </span>
            <span
              class="ant-menu-title-content"
            >
              <a
                href="/project/default/overview"
              >
                menu.projectOverview
              </a>
            </span>
          </li>
          <li
            class="ant-menu-submenu ant-menu-submenu-inline ant-menu-submenu-open"
            role="none"
          >
            <div
              aria-controls="rc-menu-uuid-test-order-popup"
              aria-expanded="true"
              aria-haspopup="true"
              class="ant-menu-submenu-title"
              data-menu-id="rc-menu-uuid-test-order"
              role="menuitem"
              style="padding-left: 24px;"
              tabindex="-1"
            >
              <span
                aria-label="console-sql"
                class="anticon anticon-console-sql ant-menu-item-icon"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="console-sql"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <defs>
                    <style />
                  </defs>
                  <path
                    d="M301.3 496.7c-23.8 0-40.2-10.5-41.6-26.9H205c.9 43.4 36.9 70.3 93.9 70.3 59.1 0 95-28.4 95-75.5 0-35.8-20-55.9-64.5-64.5l-29.1-5.6c-23.8-4.7-33.8-11.9-33.8-24.2 0-15 13.3-24.5 33.4-24.5 20.1 0 35.3 11.1 36.6 27h53c-.9-41.7-37.5-70.3-90.3-70.3-54.4 0-89.7 28.9-89.7 73 0 35.5 21.2 58 62.5 65.8l29.7 5.9c25.8 5.2 35.6 11.9 35.6 24.4.1 14.7-14.5 25.1-36 25.1z"
                  />
                  <path
                    d="M928 140H96c-17.7 0-32 14.3-32 32v496c0 17.7 14.3 32 32 32h380v112H304c-8.8 0-16 7.2-16 16v48c0 4.4 3.6 8 8 8h432c4.4 0 8-3.6 8-8v-48c0-8.8-7.2-16-16-16H548V700h380c17.7 0 32-14.3 32-32V172c0-17.7-14.3-32-32-32zm-40 488H136V212h752v416z"
                  />
                  <path
                    d="M828.5 486.7h-95.8V308.5h-57.4V534h153.2zm-298.6 53.4c14.1 0 27.2-2 39.1-5.8l13.3 20.3h53.3L607.9 511c21.1-20 33-51.1 33-89.8 0-73.3-43.3-118.8-110.9-118.8s-111.2 45.3-111.2 118.8c-.1 73.7 43 118.9 111.1 118.9zm0-190c31.6 0 52.7 27.7 52.7 71.1 0 16.7-3.6 30.6-10 40.5l-5.2-6.9h-48.8L542 491c-3.9.9-8 1.4-12.2 1.4-31.7 0-52.8-27.5-52.8-71.2.1-43.6 21.2-71.1 52.9-71.1z"
                  />
                </svg>
              </span>
              <span
                class="ant-menu-title-content"
              >
                menu.order
              </span>
              <i
                class="ant-menu-submenu-arrow"
              />
            </div>
            <ul
              class="ant-menu ant-menu-sub ant-menu-inline"
              data-menu-list="true"
              id="rc-menu-uuid-test-order-popup"
            >
              <li
                class="ant-menu-item"
                data-menu-id="rc-menu-uuid-test-orderList"
                role="menuitem"
                style="padding-left: 48px;"
                tabindex="-1"
              >
                <span
                  aria-label="bars"
                  class="anticon anticon-bars ant-menu-item-icon"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="bars"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="0 0 1024 1024"
                    width="1em"
                  >
                    <path
                      d="M912 192H328c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 284H328c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 284H328c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h584c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM104 228a56 56 0 10112 0 56 56 0 10-112 0zm0 284a56 56 0 10112 0 56 56 0 10-112 0zm0 284a56 56 0 10112 0 56 56 0 10-112 0z"
                    />
                  </svg>
                </span>
                <span
                  class="ant-menu-title-content"
                >
                  <a
                    href="/project/default/order"
                  >
                    menu.orderList
                  </a>
                </span>
              </li>
            </ul>
          </li>
          <li
            class="ant-menu-submenu ant-menu-submenu-inline"
            role="none"
          >
            <div
              aria-controls="rc-menu-uuid-test-plane-popup"
              aria-expanded="false"
              aria-haspopup="true"
              class="ant-menu-submenu-title"
              data-menu-id="rc-menu-uuid-test-plane"
              role="menuitem"
              style="padding-left: 24px;"
              tabindex="-1"
            >
              <span
                aria-label="ci-circle"
                class="anticon anticon-ci-circle ant-menu-item-icon"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="ci-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372zm218-572.1h-50.4c-4.4 0-8 3.6-8 8v384.2c0 4.4 3.6 8 8 8H730c4.4 0 8-3.6 8-8V319.9c0-4.4-3.6-8-8-8zm-281.4 49.6c49.5 0 83.1 31.5 87 77.6.4 4.2 3.8 7.4 8 7.4h52.6c2.4 0 4.4-2 4.4-4.4 0-81.2-64-138.1-152.3-138.1C345.4 304 286 373.5 286 488.4v49c0 114 59.4 182.6 162.3 182.6 88 0 152.3-55.1 152.3-132.5 0-2.4-2-4.4-4.4-4.4h-52.7c-4.2 0-7.6 3.2-8 7.3-4.2 43-37.7 72.4-87 72.4-61.1 0-95.6-44.9-95.6-125.2v-49.3c.1-81.4 34.6-126.8 95.7-126.8z"
                  />
                </svg>
              </span>
              <span
                class="ant-menu-title-content"
              >
                menu.auditPlane
              </span>
              <i
                class="ant-menu-submenu-arrow"
              />
            </div>
          </li>
          <li
            class="ant-menu-item"
            data-menu-id="rc-menu-uuid-test-data"
            role="menuitem"
            style="padding-left: 24px;"
            tabindex="-1"
          >
            <span
              aria-label="database"
              class="anticon anticon-database ant-menu-item-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="database"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M832 64H192c-17.7 0-32 14.3-32 32v832c0 17.7 14.3 32 32 32h640c17.7 0 32-14.3 32-32V96c0-17.7-14.3-32-32-32zm-600 72h560v208H232V136zm560 480H232V408h560v208zm0 272H232V680h560v208zM304 240a40 40 0 1080 0 40 40 0 10-80 0zm0 272a40 40 0 1080 0 40 40 0 10-80 0zm0 272a40 40 0 1080 0 40 40 0 10-80 0z"
                />
              </svg>
            </span>
            <span
              class="ant-menu-title-content"
            >
              <a
                href="/project/default/data"
              >
                menu.dataSource
              </a>
            </span>
          </li>
          <li
            class="ant-menu-item"
            data-menu-id="rc-menu-uuid-test-member"
            role="menuitem"
            style="padding-left: 24px;"
            tabindex="-1"
          >
            <span
              aria-label="user"
              class="anticon anticon-user ant-menu-item-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="user"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M858.5 763.6a374 374 0 00-80.6-119.5 375.63 375.63 0 00-119.5-80.6c-.4-.2-.8-.3-1.2-.5C719.5 518 760 444.7 760 362c0-137-111-248-248-248S264 225 264 362c0 82.7 40.5 156 102.8 201.1-.4.2-.8.3-1.2.5-44.8 18.9-85 46-119.5 80.6a375.63 375.63 0 00-80.6 119.5A371.7 371.7 0 00136 901.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8 2-77.2 33-149.5 87.8-204.3 56.7-56.7 132-87.9 212.2-87.9s155.5 31.2 212.2 87.9C779 752.7 810 825 812 902.2c.1 4.4 3.6 7.8 8 7.8h60a8 8 0 008-8.2c-1-47.8-10.9-94.3-29.5-138.2zM512 534c-45.9 0-89.1-17.9-121.6-50.4S340 407.9 340 362c0-45.9 17.9-89.1 50.4-121.6S466.1 190 512 190s89.1 17.9 121.6 50.4S684 316.1 684 362c0 45.9-17.9 89.1-50.4 121.6S557.9 534 512 534z"
                />
              </svg>
            </span>
            <span
              class="ant-menu-title-content"
            >
              <a
                href="/project/default/member"
              >
                menu.member
              </a>
            </span>
          </li>
          <li
            class="ant-menu-item"
            data-menu-id="rc-menu-uuid-test-ruleTemplate"
            role="menuitem"
            style="padding-left: 24px;"
            tabindex="-1"
          >
            <span
              aria-label="audit"
              class="anticon anticon-audit ant-menu-item-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="audit"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M296 250c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h384c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H296zm184 144H296c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8h184c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8zm-48 458H208V148h560v320c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V108c0-17.7-14.3-32-32-32H168c-17.7 0-32 14.3-32 32v784c0 17.7 14.3 32 32 32h264c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm440-88H728v-36.6c46.3-13.8 80-56.6 80-107.4 0-61.9-50.1-112-112-112s-112 50.1-112 112c0 50.7 33.7 93.6 80 107.4V764H520c-8.8 0-16 7.2-16 16v152c0 8.8 7.2 16 16 16h352c8.8 0 16-7.2 16-16V780c0-8.8-7.2-16-16-16zM646 620c0-27.6 22.4-50 50-50s50 22.4 50 50-22.4 50-50 50-50-22.4-50-50zm180 266H566v-60h260v60z"
                />
              </svg>
            </span>
            <span
              class="ant-menu-title-content"
            >
              <a
                href="/project/default/rule/template"
              >
                menu.ruleTemplate
              </a>
            </span>
          </li>
          <li
            class="ant-menu-item"
            data-menu-id="rc-menu-uuid-test-progress"
            role="menuitem"
            style="padding-left: 24px;"
            tabindex="-1"
          >
            <span
              aria-label="node-index"
              class="anticon anticon-node-index ant-menu-item-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="node-index"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <defs>
                  <style />
                </defs>
                <path
                  d="M843.5 737.4c-12.4-75.2-79.2-129.1-155.3-125.4S550.9 676 546 752c-153.5-4.8-208-40.7-199.1-113.7 3.3-27.3 19.8-41.9 50.1-49 18.4-4.3 38.8-4.9 57.3-3.2 1.7.2 3.5.3 5.2.5 11.3 2.7 22.8 5 34.3 6.8 34.1 5.6 68.8 8.4 101.8 6.6 92.8-5 156-45.9 159.2-132.7 3.1-84.1-54.7-143.7-147.9-183.6-29.9-12.8-61.6-22.7-93.3-30.2-14.3-3.4-26.3-5.7-35.2-7.2-7.9-75.9-71.5-133.8-147.8-134.4-76.3-.6-140.9 56.1-150.1 131.9s40 146.3 114.2 163.9c74.2 17.6 149.9-23.3 175.7-95.1 9.4 1.7 18.7 3.6 28 5.8 28.2 6.6 56.4 15.4 82.4 26.6 70.7 30.2 109.3 70.1 107.5 119.9-1.6 44.6-33.6 65.2-96.2 68.6-27.5 1.5-57.6-.9-87.3-5.8-8.3-1.4-15.9-2.8-22.6-4.3-3.9-.8-6.6-1.5-7.8-1.8l-3.1-.6c-2.2-.3-5.9-.8-10.7-1.3-25-2.3-52.1-1.5-78.5 4.6-55.2 12.9-93.9 47.2-101.1 105.8-15.7 126.2 78.6 184.7 276 188.9 29.1 70.4 106.4 107.9 179.6 87 73.3-20.9 119.3-93.4 106.9-168.6zM329.1 345.2a83.3 83.3 0 11.01-166.61 83.3 83.3 0 01-.01 166.61zM695.6 845a83.3 83.3 0 11.01-166.61A83.3 83.3 0 01695.6 845z"
                />
              </svg>
            </span>
            <span
              class="ant-menu-title-content"
            >
              <a
                href="/project/default/progress"
              >
                menu.progressManage
              </a>
            </span>
          </li>
          <li
            class="ant-menu-item"
            data-menu-id="rc-menu-uuid-test-Whitelist"
            role="menuitem"
            style="padding-left: 24px;"
            tabindex="-1"
          >
            <span
              aria-label="profile"
              class="anticon anticon-profile ant-menu-item-icon"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="profile"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656zM492 400h184c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H492c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8zm0 144h184c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H492c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8zm0 144h184c4.4 0 8-3.6 8-8v-48c0-4.4-3.6-8-8-8H492c-4.4 0-8 3.6-8 8v48c0 4.4 3.6 8 8 8zM340 368a40 40 0 1080 0 40 40 0 10-80 0zm0 144a40 40 0 1080 0 40 40 0 10-80 0zm0 144a40 40 0 1080 0 40 40 0 10-80 0z"
                />
              </svg>
            </span>
            <span
              class="ant-menu-title-content"
            >
              <a
                href="/project/default/whitelist"
              >
                menu.whitelist
              </a>
            </span>
          </li>
        </ul>
        <div
          aria-hidden="true"
          style="display: none;"
        />
      </div>
    </aside>
    <main
      class="ant-layout-content"
    >
      <article>
        <div
          class="ant-page-header"
        >
          <div
            class="ant-page-header-heading"
          >
            <div
              class="ant-page-header-heading-left"
            >
              <span
                class="ant-page-header-heading-title"
                title="projectManage.projectOverview.pageTitle"
              >
                projectManage.projectOverview.pageTitle
              </span>
            </div>
            <span
              class="ant-page-header-heading-extra"
            >
              <button
                class="ant-btn"
                data-testid="refresh-project-info"
                type="button"
              >
                <span
                  aria-label="sync"
                  class="anticon anticon-sync anticon-spin"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="sync"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M168 504.2c1-43.7 10-86.1 26.9-126 17.3-41 42.1-77.7 73.7-109.4S337 212.3 378 195c42.4-17.9 87.4-27 133.9-27s91.5 9.1 133.8 27A341.5 341.5 0 01755 268.8c9.9 9.9 19.2 20.4 27.8 31.4l-60.2 47a8 8 0 003 14.1l175.7 43c5 1.2 9.9-2.6 9.9-7.7l.8-180.9c0-6.7-7.7-10.5-12.9-6.3l-56.4 44.1C765.8 155.1 646.2 92 511.8 92 282.7 92 96.3 275.6 92 503.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8zm756 7.8h-60c-4.4 0-7.9 3.5-8 7.8-1 43.7-10 86.1-26.9 126-17.3 41-42.1 77.8-73.7 109.4A342.45 342.45 0 01512.1 856a342.24 342.24 0 01-243.2-100.8c-9.9-9.9-19.2-20.4-27.8-31.4l60.2-47a8 8 0 00-3-14.1l-175.7-43c-5-1.2-9.9 2.6-9.9 7.7l-.7 181c0 6.7 7.7 10.5 12.9 6.3l56.4-44.1C258.2 868.9 377.8 932 512.2 932c229.2 0 415.5-183.7 419.8-411.8a8 8 0 00-8-8.2z"
                    />
                  </svg>
                </span>
              </button>
            </span>
          </div>
          <div
            class="ant-page-header-content"
          />
        </div>
        <section
          class="padding-content"
        >
          <div
            class="ant-spin ant-spin-spinning"
          >
            <span
              class="ant-spin-dot ant-spin-dot-spin"
            >
              <i
                class="ant-spin-dot-item"
              />
              <i
                class="ant-spin-dot-item"
              />
              <i
                class="ant-spin-dot-item"
              />
              <i
                class="ant-spin-dot-item"
              />
            </span>
          </div>
        </section>
      </article>
    </main>
  </section>
</div>
`;
